filament/src/materials/ssao.mat

```diff
@@ -117,15 +117,30 @@ fragment {
         return (d >= -bias ? 0.0 : rangeCheck);
     }
 
+    float square(in float x) {
+        return x * x;
+    }
+
     void material(inout MaterialInputs material) {
         prepareMaterial(material);
 
         vec2 uv = variable_vertex.xy; // interpolated to pixel center
         vec3 noise = getNoise(uv);
         float depth = sampleDepthLinear(uv);
         vec3 origin = computeViewSpacePositionFromDepth(variable_vertex.zw, depth);
-        vec3 normal = computeViewSpaceNormal(origin);
+        vec3 normal = computeViewSpaceNormalNotNormalized(origin);
+
+        // attempt to reject "bad" normals that were reconstructed at an edge that cause
+        // false occlusion (black spots)
+        // Normal's length should be small before they're normalized, unless they're bad ones.
+        if (dot(normal, normal) >= (square(origin.z * origin.z * 0.000061))) {
```

Comment:
Any chance we could use a constant here instead of `square(origin.z…)`?

Reply:
unfortunately I don't think so -- z^2 is the expected length of the normal, the test is essentially equivalent to : length(normal) / z^2 >= 2^-14 -- that is we test if the normal is 2^14 larger than expected.